### PR TITLE
Adversarial bias mitigation

### PR DESCRIPTION
**Pull request by ArjunSubramonian**
Changes/additions proposed in this pull request:

- Added `AdversarialBiasMitigator`, a Model wrapper to adversarially mitigate biases in predictions produced by a pretrained model for a downstream task. Tests are in: https://github.com/allenai/allennlp-models/pull/281/.
- Fixed `IndexOutOfBoundsException` in `MultiOptimizer` when checking if optimizer received any parameters.
- Changed behavior of `MultiOptimizer` so that while a default optimizer is still required, an error is not thrown if the default optimizer receives no parameters.

Depends on https://github.com/allenai/allennlp-models/pull/281/.

### CODE REVIEW

**Comment by dirkgr** [comment]
I have misspelled "adversarial" all my life 🤦🏻 . Seems like autocorrect bailed me out a lot.

File: allennlp/training/optimizers.py [hunk 1]
```diff
@@ -278,6 +278,11 @@ def __init__(
                     " Alternatively, you can remove this optimizer from the provided `optimizers`"
                     " if it is not relevant to a particular parameter group."
                 )
+        # default optimizer is required, but may not be used
```
Comment (dirkgr) [inline allennlp/training/optimizers.py:281]: "must not be used" or "using it is optional"?

Comment (dirkgr) [reply allennlp/training/optimizers.py:281]: I know what you mean from context, but it would be easier to read if worded unambiguously.

**Review by dirkgr — APPROVED** [review]

File: allennlp/fairness/adversarial_bias_mitigator.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+        batch_outputs["adversary_loss"].backward(retain_graph=True)
+        # trainer.model is expected to have `predictor` and `adversary` data members
+        adversary_loss_grad = {
+            name: param.grad.clone()
+            for name, param in trainer.model.predictor.named_parameters()
+            if param.grad is not None
+        }
+
+        trainer.model.predictor.zero_grad()
+        batch_outputs["loss"].backward()
```
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/fairness/adversarial_bias_mitigator.py:284]: That's a really good point about part of the computation graph not getting erased! Upon further reading, it looks like the computation graph will stay around until `adversary_loss` goes out of scope. So I added this in the callback to remove all references to the `adversary_loss` in the graph and instead keep a view of the loss that's not in the graph:
```
# remove adversary_loss from computation graph
batch_outputs["adversary_loss"] = batch_outputs["adversary_loss"].detach()
```

**Review by dirkgr — APPROVED** [review]

File: allennlp/fairness/adversarial_bias_mitigator.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+(https://api.semanticscholar.org/CorpusID:9424845).
+Proceedings of the 2018 AAAI/ACM Conference on AI, Ethics, and Society.
+and [Mitigating Unwanted Biases in Word Embeddings
+with Adversarial Learning](https://colab.research.google.com/notebooks/
+ml_fairness/adversarial_debiasing.ipynb) colab notebook.
+
+Adversarial networks mitigate some biases based on the idea that
+predicting an outcome Y given an input X should ideally be independent
+of some protected variable Z. Informally, "knowing Y would not help
+you predict Z any better than chance" (Zaldivar et al., 2018). This
```
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/fairness/adversarial_bias_mitigator.py:17]: The latter, it's stating that knowing the outcome shouldn't give you information about the protected variable.

File: allennlp/fairness/adversarial_bias_mitigator.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+@Model.register("adversarial_bias_mitigator")
+class AdversarialBiasMitigator(Model):
+    """
+    Wrapper class to adversarially mitigate biases in any pretrained Model.
+
+    # Parameters
+
+    vocab : `Vocabulary`
+        Vocabulary of predictor.
+    predictor : `Model`
```
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/fairness/adversarial_bias_mitigator.py:69]: Yes, I'm adhering to the paper's terminology.
